File: examples/setup_mongodb_sample.py
#!/usr/bin/env python3
"""
MongoDB Sample Database Setup Script
Creates a sample database with 10GB of user data
"""

import random
import string
from datetime import datetime, timedelta
from pymongo import MongoClient
import time

# MongoDB connection
MONGO_URI = "mongodb://admin:password123@localhost:27017/"
DATABASE_NAME = "sample"
COLLECTION_NAME = "user"
TARGET_SIZE_GB = 10
BATCH_SIZE = 500

def random_string(length):
    """Generate a random string of specified length"""
    return ''.join(random.choices(string.ascii_letters + string.digits, k=length))

def random_date(start_days_ago=365):
    """Generate a random date within the last N days"""
    days_ago = random.randint(0, start_days_ago)
    return datetime.now() - timedelta(days=days_ago)

def generate_data_array(size):
    """Generate an array of data objects"""
    return [
        {
            "key": random_string(20),
            "value": random_string(100),
            "timestamp": random_date(),
            "metadata": random_string(50)
        }
        for _ in range(size)
    ]

def generate_user(user_id):
    """Generate a single user document with lots of data"""
    return {
        "_id": user_id,
        "username": f"user_{user_id}",
        "email": f"user{user_id}@example.com",
        "firstName": random_string(15),
        "lastName": random_string(20),
        "age": random.randint(18, 78),
        "createdAt": random_date(365),
        "updatedAt": datetime.now(),
        "address": {
            "street": random_string(30),
            "city": random_string(20),
            "state": random_string(15),
            "zipCode": random_string(10),
            "country": random_string(20),
            "coordinates": {
                "lat": random.uniform(-90, 90),
                "lng": random.uniform(-180, 180)
            }
        },
        "profile": {
            "bio": random_string(500),
            "interests": [random_string(25) for _ in range(20)],
            "skills": [random_string(20) for _ in range(15)],
            "education": [
                {
                    "institution": random_string(40),
                    "degree": random_string(30),
                    "year": random.randint(1990, 2024),
                    "description": random_string(200)
                }
                for _ in range(5)
            ],
            "experience": [
                {
                    "company": random_string(35),
                    "position": random_string(30),
                    "duration": random_string(20),
                    "description": random_string(300)
                }
                for _ in range(8)
            ]
        },
        "preferences": {
            "language": random_string(10),
            "timezone": random_string(20),
            "notifications": {
                "email": random.choice([True, False]),
                "push": random.choice([True, False]),
                "sms": random.choice([True, False])
            },
            "settings": generate_data_array(10)
        },
        "activityLog": generate_data_array(30),
        "transactions": generate_data_array(25),
        "messages": [
            {
                "id": random_string(24),
                "from": random_string(20),
                "to": random_string(20),
                "subject": random_string(50),
                "body": random_string(500),
                "timestamp": random_date(30),
                "read": random.choice([True, False]),
                "attachments": [
                    {
                        "name": random_string(30),
                        "size": random.randint(1000, 1000000),
                        "type": random_string(15)
                    }
                    for _ in range(3)
                ]
            }
            for _ in range(50)
        ],
        "socialMedia": {
            "posts": generate_data_array(40),
            "connections": [random_string(24) for _ in range(100)],
            "groups": [
                {
                    "id": random_string(24),
                    "name": random_string(30),
                    "joined": random_date(365)
                }
                for _ in range(20)
            ]
        },
        "largeData": {
            "blob1": random_string(5000),
            "blob2": random_string(5000),
            "blob3": random_string(5000),
            "blob4": random_string(5000),
            "blob5": random_string(5000)
        }
    }

def get_collection_size(collection):
    """Get the current size of the collection in bytes"""
    stats = collection.database.command("collStats", collection.name)
    return stats.get("size", 0)

def main():
    print("=" * 50)
    print("MongoDB Sample Database Setup")
    print("=" * 50)
    print(f"\nTarget size: {TARGET_SIZE_GB} GB")
    print(f"Database: {DATABASE_NAME}")
    print(f"Collection: {COLLECTION_NAME}")
    print(f"Batch size: {BATCH_SIZE} documents")
    print("\nConnecting to MongoDB...")
    
    try:
        # Connect to MongoDB
        client = MongoClient(MONGO_URI, serverSelectionTimeoutMS=5000)
        client.server_info()
        print("✓ Connected to MongoDB successfully")
        
        # Get database and collection
        db = client[DATABASE_NAME]
        collection = db[COLLECTION_NAME]
        
        # Drop existing collection
        print(f"\nDropping existing '{COLLECTION_NAME}' collection if it exists...")
        collection.drop()
        print("✓ Collection dropped")
        
        # Generate and insert data
        target_size_bytes = TARGET_SIZE_GB * 1024 * 1024 * 1024
        total_inserted = 0
        start_time = time.time()
        
        print(f"\nStarting data generation (target: {TARGET_SIZE_GB} GB)...")
        print("This will take approximately 10-15 minutes...\n")
        
        while True:
            # Check current size
            current_size = get_collection_size(collection)
            
            if current_size >= target_size_bytes:
                break
            
            # Generate batch of documents
            batch = []
            for i in range(BATCH_SIZE):
                user_id = total_inserted + i + 1
                batch.append(generate_user(user_id))
            
            # Insert batch
            try:
                collection.insert_many(batch, ordered=False)
                total_inserted += BATCH_SIZE
                
                # Progress update
                progress = (current_size / target_size_bytes) * 100
                size_gb = current_size / (1024 * 1024 * 1024)
                elapsed = time.time() - start_time
                docs_per_sec = total_inserted / elapsed if elapsed > 0 else 0
                
                print(f"Progress: {progress:5.2f}% | "
                      f"Documents: {total_inserted:,} | "
                      f"Size: {size_gb:.3f} GB | "
                      f"Speed: {docs_per_sec:.0f} docs/sec")
                
            except Exception as e:
                print(f"Error inserting batch: {e}")
                continue
        
        # Create indexes
        print("\n\nCreating indexes...")
        collection.create_index("username")
        collection.create_index("email")
        collection.create_index("createdAt")
        collection.create_index("address.city")
        collection.create_index("age")
        print("✓ Indexes created")
        
        # Final statistics
        final_size = get_collection_size(collection)
        doc_count = collection.count_documents({})
        stats = db.command("collStats", COLLECTION_NAME)
        
        elapsed_time = time.time() - start_time
        
        print("\n" + "=" * 50)
        print("Setup Complete!")
        print("=" * 50)
        print(f"Total documents:       {doc_count:,}")
        print(f"Collection size:       {final_size / (1024**3):.3f} GB")
        print(f"Storage size:          {stats.get('storageSize', 0) / (1024**3):.3f} GB")
        print(f"Average document size: {stats.get('avgObjSize', 0) / 1024:.2f} KB")
        print(f"Number of indexes:     {stats.get('nindexes', 0)}")
        print(f"Time elapsed:          {elapsed_time / 60:.2f} minutes")
        print(f"\nConnection string: {MONGO_URI}{DATABASE_NAME}")
        
    except Exception as e:
        print(f"\n✗ Error: {e}")
        return 1
    
    finally:
        if 'client' in locals():
            client.close()
    
    return 0

if __name__ == "__main__":
    exit(main())

File: examples/setup_mongodb_sample.js
// MongoDB script to create sample database with 10GB of user data
// Run with: docker exec -i mongodb-dev mongosh -u admin -p password123 --authenticationDatabase admin < setup_mongodb_sample.js

// Switch to sample database
use sample;

print("Starting to generate 10GB of user data...");

// Drop existing collection if it exists
db.user.drop();

// Create the user collection
db.createCollection("user");

// Function to generate random user data
function generateUser(id) {
    // Generate random string data to increase document size
    const generateRandomString = (length) => {
        const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let result = '';
        for (let i = 0; i < length; i++) {
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    };

    // Generate random data arrays to increase size
    const generateDataArray = (size) => {
        const arr = [];
        for (let i = 0; i < size; i++) {
            arr.push({
                key: generateRandomString(20),
                value: generateRandomString(100),
                timestamp: new Date(Date.now() - Math.floor(Math.random() * 365 * 24 * 60 * 60 * 1000)),
                metadata: generateRandomString(50)
            });
        }
        return arr;
    };

    return {
        _id: id,
        username: `user_${id}`,
        email: `user${id}@example.com`,
        firstName: generateRandomString(15),
        lastName: generateRandomString(20),
        age: Math.floor(Math.random() * 60) + 18,
        createdAt: new Date(Date.now() - Math.floor(Math.random() * 365 * 24 * 60 * 60 * 1000)),
        updatedAt: new Date(),
        address: {
            street: generateRandomString(30),
            city: generateRandomString(20),
            state: generateRandomString(15),
            zipCode: generateRandomString(10),
            country: generateRandomString(20),
            coordinates: {
                lat: Math.random() * 180 - 90,
                lng: Math.random() * 360 - 180
            }
        },
        profile: {
            bio: generateRandomString(500),
            interests: Array(20).fill(null).map(() => generateRandomString(25)),
            skills: Array(15).fill(null).map(() => generateRandomString(20)),
            education: Array(5).fill(null).map(() => ({
                institution: generateRandomString(40),
                degree: generateRandomString(30),
                year: 1990 + Math.floor(Math.random() * 30),
                description: generateRandomString(200)
            })),
            experience: Array(8).fill(null).map(() => ({
                company: generateRandomString(35),
                position: generateRandomString(30),
                duration: generateRandomString(20),
                description: generateRandomString(300)
            }))
        },
        preferences: {
            language: generateRandomString(10),
            timezone: generateRandomString(20),
            notifications: {
                email: Math.random() > 0.5,
                push: Math.random() > 0.5,
                sms: Math.random() > 0.5
            },
            settings: generateDataArray(10)
        },
        activityLog: generateDataArray(30),
        transactions: generateDataArray(25),
        messages: Array(50).fill(null).map(() => ({
            id: generateRandomString(24),
            from: generateRandomString(20),
            to: generateRandomString(20),
            subject: generateRandomString(50),
            body: generateRandomString(500),
            timestamp: new Date(Date.now() - Math.floor(Math.random() * 30 * 24 * 60 * 60 * 1000)),
            read: Math.random() > 0.5,
            attachments: Array(3).fill(null).map(() => ({
                name: generateRandomString(30),
                size: Math.floor(Math.random() * 1000000),
                type: generateRandomString(15)
            }))
        })),
        socialMedia: {
            posts: generateDataArray(40),
            connections: Array(100).fill(null).map(() => generateRandomString(24)),
            groups: Array(20).fill(null).map(() => ({
                id: generateRandomString(24),
                name: generateRandomString(30),
                joined: new Date(Date.now() - Math.floor(Math.random() * 365 * 24 * 60 * 60 * 1000))
            }))
        },
        largeData: {
            blob1: generateRandomString(5000),
            blob2: generateRandomString(5000),
            blob3: generateRandomString(5000),
            blob4: generateRandomString(5000),
            blob5: generateRandomString(5000)
        }
    };
}

// Batch insert for better performance
const batchSize = 1000;
let totalInserted = 0;
let totalSize = 0;
const targetSize = 10 * 1024 * 1024 * 1024; // 10GB in bytes

print(`Target size: ${(targetSize / 1024 / 1024 / 1024).toFixed(2)} GB`);

while (totalSize < targetSize) {
    const batch = [];
    const startId = totalInserted + 1;

    // Generate a batch of documents
    for (let i = 0; i < batchSize; i++) {
        batch.push(generateUser(startId + i));
    }

    // Insert the batch
    try {
        db.user.insertMany(batch, { ordered: false });
        totalInserted += batchSize;

        // Check current collection size
        const stats = db.user.stats();
        totalSize = stats.size;

        const progress = (totalSize / targetSize * 100).toFixed(2);
        const sizeGB = (totalSize / 1024 / 1024 / 1024).toFixed(3);

        print(`Progress: ${progress}% | Documents: ${totalInserted} | Size: ${sizeGB} GB`);

        // Add a small delay to avoid overwhelming the system
        sleep(100);
    } catch (e) {
        print(`Error inserting batch: ${e}`);
        break;
    }
}

// Create indexes for better query performance
print("\nCreating indexes...");
db.user.createIndex({ username: 1 });
db.user.createIndex({ email: 1 });
db.user.createIndex({ createdAt: -1 });
db.user.createIndex({ "address.city": 1 });
db.user.createIndex({ age: 1 });

// Final statistics
const finalStats = db.user.stats();
print("\n=== Final Statistics ===");
print(`Total documents: ${db.user.countDocuments()}`);
print(`Collection size: ${(finalStats.size / 1024 / 1024 / 1024).toFixed(3)} GB`);
print(`Storage size: ${(finalStats.storageSize / 1024 / 1024 / 1024).toFixed(3)} GB`);
print(`Average document size: ${(finalStats.avgObjSize / 1024).toFixed(2)} KB`);
print(`Indexes: ${finalStats.nindexes}`);
print("\nDatabase setup complete!");
